core/src/main/java/com/adobe/aem/modernize/component/impl/ComponentTreeRewriter.java

```diff
@@ -1,116 +1,115 @@
-
-        long tick = System.currentTimeMillis();
-        Session session = root.getSession();
-        boolean success = false;
-        Node startNode = root;
-        try {
-            Node parent = root.getParent();
-
-            boolean ordered = parent.getPrimaryNodeType().hasOrderableChildNodes();
-            NodeIterator siblings = parent.getNodes();
-            while (siblings.hasNext()) {
-                Node node = siblings.nextNode();
-
-                // If this is the node to update.
-                if (startNode.getPath().equals(node.getPath())) {
-
-                    // Apply the matching rule.
-                    for (ComponentRewriteRule rule : rules) {
-                        // check for a match
-                        if (rule.matches(node)) {
-                            logger.debug("Rule {} matched subtree rooted at {}", rule, node.getPath());
-                            // the rule matched, rewrite the tree
-                            node = rule.applyTo(node, new HashSet<>());
-                            startNode = node;
-                            break;
-                        }
-                    }
-                }
-
-                // Keep the parent in order as we are updating one of it's children.
-                if (ordered) {
-                    parent.orderBefore(node.getName(), null);
-                }
-            }
-            success = true;
-        } finally {
-            if (!success) {
-                // an exception has been thrown: try to revert changes
-                try {
-                    session.refresh(false);
-                } catch (RepositoryException e) {
-                    logger.warn("Could not revert changes", e);
-                }
-            }
+  private static final Logger logger = LoggerFactory.getLogger(ComponentTreeRewriter.class);
+
+  /**
+   * Rewrites the specified tree according to the provided set of rules. Rules are applied according to {@link TreeTraverser} order.
+   *
+   * Changes are not saved.
+   *
+   * An exception is thrown if any error occurs terminating the rewrite at that location in the traversal. Changes are not reverted.
+   *
+   * @param root The root of the tree to be rewritten
+   * @param rules The list of rules to apply to the tree
+   * @return the root node of the rewritten tree, or null if it was removed
+   * @throws RewriteException if the rewrite operation fails
+   * @throws RepositoryException if there is a problem with the repository
+   */
+  @Nullable
+  static Node rewrite(@NotNull Node root, @NotNull List<RewriteRule> rules) throws RewriteException, RepositoryException {
+    String rootPath = root.getPath();
+    logger.debug("Rewriting content tree rooted at: {}", rootPath);
+    long tick = System.currentTimeMillis();
+    Node startNode = root;
+    boolean matched;
+
+    Set<String> finalPaths = new LinkedHashSet<>();
+
+    do {
+      matched = false;
```

Comment:
There's a lot happening inside this loop, consider refactoring into 1/2 methods so that you can ensure the logic is sound